Add EIP-5539: Revocation List Registry (#5539)
ethereum/EIPs: EIPS/eip-template.md

---

> eip: <to be assigned>

Reviewer: ```suggestion
eip: 5539
```
Don't forget to rename the file too!

title: Revocation List Registry
description: Registry of revocation lists for revoking arbitrary data.
author: Philipp Bolte (@strumswell), Lauritz Leifermann (@lleifermann), Dennis von der Bey (@DennisVonDerBey)

> discussions-to: <URL>

Reviewer: Please create a thread on https://ethereum-magicians.org/

Author: Created a post there but it needs to be unblocked by a moderator first.

Reviewer: Huh, that's weird and shouldn't happen. CC @jpitts

Author: To be fair i never had an account there so i may have triggered some spam system by immediately posting 😄. Here's the account -> https://ethereum-magicians.org/u/lleifermann/

status: Draft
type: Standards Track
category: ERC
created: 2022-08-26

> requires (*optional): EIP-712

Reviewer: ```suggestion
requires: 712
```

> requires (*optional): EIP-712

Reviewer: ```suggestion
requires: 712
```
Required: This is the correct format for the requires header.

function batchChangeStatuses(bool[] revokedStatuses, address[] namespaces, bytes32[] lists, bytes32[] keys) public;
```

**batchChangeStatusesSigned** ([see meta transactions](#Meta-transactions))
**OPTIONAL** implements a function to change multiple revocation statuses in different revocation lists and namespaces at once with a raw signature.
```solidity
function batchChangeStatusesSigned(bool[] revokedStatuses, address[] namespaces, bytes32[] lists, bytes32[] keys, uint8 sigV, bytes32 sigR, bytes32 sigS) public;
```

**batchChangeListStatuses**
**OPTIONAL** implements a function to change multiple revocation statuses in a specific namespace's revocation list.
```solidity
function batchChangeListStatuses(bool[] revokedStatuses, address namespace, bytes32 list, bytes32[] keys) public;
```

**batchChangeListStatusesSigned** ([see meta transactions](#Meta-transactions))
**OPTIONAL** implements a function to change multiple revocation statuses in a specific namespace's revocation list with a raw signature.
```solidity
function batchChangeListStatusesSigned(bool[] revokedStatuses, address namespace, bytes32 list, bytes32[] keys, uint8 sigV, bytes32 sigR, bytes32 sigS) public;
```

> ### Owner managment

Reviewer: ```suggestion
### Owner management
```

> ### Owner managment

Reviewer: ```suggestion
### Owner management
```
Required: Typo

**MUST** be emitted when `addListDelegate` or `addListDelegateSigned` was successfully executed.

```solidity
event ListDelegateAdded(
    address indexed namespace,
    address indexed delegate,
    bytes32 indexed list
);
```

**DelegateRemoved**
**MUST** be emitted when `removeListDelegate` or `removeListDelegateSigned` was successfully executed.

```solidity
event ListDelegateRemoved(
    address indexed namespace,
    address indexed delegate,
    bytes32 indexed list
);
```

### Meta transactions

This section uses the following terms: 

**`transaction signer`:** An Ethereum address that signs arbitrary data for the contract to execute **BUT** does not commit the transaction. 
**`transaction sender`**: An Ethereum address that takes signed data from a **transaction signer** and commits it wrapped with its own signature to the smart contract.

An address (**transaction signer**) **MAY** be able to deliver a signed payload off-band to another address (**transaction sender**) that initiates the Ethereum interaction with the smart contract. The signed payload **MUST** be limited to be used only once ([Signed Hash](#Signed-Hash) + [nonces](#Nonce)).

#### Signed Hash

> The signature of the **transaction signer** **MUST** conform [EIP-712](/EIPS/eip-712), in its state that is proposed in [`ethereum/EIPs/issues/5475`](https://github.com/ethereum/EIPs/issues/5475). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.

Reviewer: ```suggestion
The signature of the **transaction signer** **MUST** conform [EIP-712](./eip-712.md). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
```
Required: External links (even to this GitHub repository) aren't allowed.
Required: Internal links must be used when applicable.

#### Nonce

This EIP **RECOMMENDS** the use of a **dedicated nonce mapping** for meta transactions. If the signature of the **transaction sender** and its meta contents are verified, the contract increases a nonce for this **transaction signer**. This effectively removes the possibility for any other sender to execute the same transaction again with another wallet. 

## Rationale

### Why the concept of namespaces?
> This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.

### Why does a namespace always represent the initial owner address? 
> The change of an owner of a list shouldn't break the link to a revocation key in it, as already existing off-chain data may depend on it. 

## Backwards Compatibility

> Not applicable

Reviewer: ```suggestion
No backward compatibility issues were found.
```
Required: Backwards Compatibility is *always* required, even if there is nothing to be compatible with.

> 
> ## Reference Implementation
> tba

Reviewer: ```suggestion
```
Required: The Reference Implementation section is optional and should be left empty when there is no content.

> ## Security Considerations
> 

Reviewer: ```suggestion
## Security Considerations
Needs discussion.

```
Required: Security Considerations is *not* an optional section.

Reviewer: I missed that. My bad.

### Meta Transactions

> The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](/EIPS/eip-712

Reviewer: ```suggestion
The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](./eip-712.md)
```

> The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](/EIPS/eip-712

Reviewer: ```suggestion
The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](./eip-712.md).
```
```suggestion
The signature of the **transaction signer** **MUST** conform [EIP-712](./eip-712.md). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
```
Required: Internal links must be used when applicable.
